Add TeeLogger

jcorbin opened the pull request:
Towards #159 
- [x] made spy logger honor level, so that I can use it to verify one useful way of using tee logger
- [x] write a basic tee logger; TeeLogger is now considered complete per all prior conversations
- [x] finish Level refactor #180 
- [x] finish CheckedMessage composition support #182

jcorbin commented on tee_logger.go:
```diff
@@ -0,0 +1,135 @@
+		Meta: ml.Meta.Clone(),
+		logs: ml.logs,
+	}
+	addFields(clone.Encoder, fields)
```
Actually, I think this is wrong, and points to the fact that I'm missing a test case for the With path (e.g. over differently encoded legs).

prashantv replied:
I think a `Clone` should be calling `With` on all the underlying loggers and creating a new `multiLogger` with the results of the `With`

jcorbin commented on tee_logger.go:
```diff
@@ -0,0 +1,135 @@
+			return nil
+		}
+	}
+	return NewCheckedMessage(ml, lvl, msg)
```
Maybe should call all sub `.Check`s and return a compound checked message.  Otherwise, this assumes that all `Logger` implementations treat level monotonically... maybe that was an intentional design limitation?

prashantv replied:
That's a good point, I think we should call `Check` on all the underlying implementations too. This becomes especially important when something like the sampled logger relies on the `Check` being called to figure out whether a log should be written

prashantv commented on tee_logger.go:
```diff
@@ -0,0 +1,135 @@
+	case 1:
+		return logs[0]
+	default:
+		lvl := logs[0].Level()
```
nit: I'd call this `minLevel`

prashantv commented on tee_logger.go:
```diff
@@ -0,0 +1,135 @@
+			}
+		}
+		ml := &multiLogger{
+			Meta: MakeMeta(nil),
```
Do we need a meta embedded? I think we're going to be implementing almost all the methods it provides ourselves so not sure if it gives us much.

jcorbin commented:
- took away the meta embedding
- simplified .With cloning, since we've got a value receiver just like Meta does
- that way when I copied the "pointer to an atomic" pattern from it, I was confident it would have same semantics as prior work
- call all the sub logger .With s
- call all the sub loggger .Check s ; the "natural" path of "a multi-logger returns a multi-checked-message" wasn't readily possible:
  - the interface return value is concrete, so I can't easily return a `type multiCheckedMessage []CheckedMessage` that apes the `.OK / .Write` contract
  - I could've tried to expand the CheckedMessage struct to support multi-ing itself..
  - ...but since it's basically "just" a `.Log`-once-guard, decided to start out by "stopping on first OK" and discarding it

Was the design choice to make `CheckedMessage` a concrete type done explicitly for performance?  That's my working suspicion since an interface value may have needed an alloc?

akshayjshah commented:
> Was the design choice to make CheckedMessage a concrete type done explicitly for performance? That's my working suspicion since an interface value may have needed an alloc?

Nailed it.

Pending changes in other PRs/as discussed, approach looks great.

jcorbin commented:
Regrettably, the "TeeLogger_Panic" isn't really as useful as it should be, since `spy.Logger` will never panic (similarly, fatal never exits).

jcorbin commented on tee_logger.go:
```diff
@@ -0,0 +1,139 @@
+// loggers. Changing the returned logger's level will change the level on all
+// wrapped loggers.
+//
+// With the exception of DFatal, the returned tee logger calls .Log(level, ...)
```
This exception is unfortunate in my view, pointing to "DFatal should be just another log level" in my opinion; we could have the exact same semantics, that `logger.DFatal` gets to make a choice, but `logger.Log` either passes the level through, or maps it to "error" always.

jcorbin commented:
Updated with an initial WIP (tests pass) of the "biggest rock" from #166: dropped `Logger.SetLevel` and `Logger.Level`, replaced with `DynamicLevel`.

I'll start breaking more incremental PRs out tomorrow.

jcorbin commented:
Ready for review.

akshayjshah commented on tee_logger.go:
```diff
@@ -0,0 +1,120 @@
+// TeeLogger creates a Logger that duplicates its log calls to two or
+// more loggers. It is similar to io.MultiWriter.
+//
+// For each logging level methad (.Debug, .Info, etc), the TeeLogger calls each
```
spelling: "method"

akshayjshah commented on tee_logger.go:
```diff
@@ -0,0 +1,120 @@
+// is returned against the TeeLogger itself. This is so that
+// tlog.Check(PanicLevel, ...).Write(...) is equivalent to tlog.Panic(...)
+// (likewise for FatalLevel).
+func TeeLogger(logs ...Logger) Logger {
```
Perhaps just `Tee`? We've got `Sample`, and we'll soon have `Sugar` and `Desugar`, all of which implicitly deal in loggers.

When we get around to teeing outputs, we can use `TeeOutput`.

jcorbin replied:
Only rub was `zap.Tee` was already taken; but I too think that `zap.Tee` should be a logger, while the "output tee" gets named differently, will update

akshayjshah commented on tee_logger.go:
```diff
@@ -0,0 +1,120 @@
+// process (using os.Exit or panic() per usual semantics).
+//
+// DFatal is handled similarly to Fatal and Panic, since it is not actually a
+// level; each sub-logger's DFatal method dynamically chooses to either cal
```
spelling: "call"

Also, I'm increasingly convinced that you're right about `DFatal` - it should be an actual level. I've filed #188 to track this.

akshayjshah commented on tee_logger.go:
```diff
@@ -0,0 +1,120 @@
+	}
+}
+
+type multiLogger []Logger
```
Awesome! Really glad that the composition ended up clean enough that we don't need any extra state at all.

akshayjshah commented on tee_logger.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+func (ml multiLogger) With(fields ...Field) Logger {
+	ml = append([]Logger(nil), ml...)
```
What's the benefit of this over `ml := make(multiLogger, len(ml))`?

akshayjshah commented on tee_logger.go:
```diff
@@ -0,0 +1,120 @@
+	for i, log := range ml {
+		ml[i] = log.With(fields...)
+	}
+	return multiLogger(ml)
```
Since `multiLogger` exposes the higher-perf `Logger` API, let's golf away the copy here. Can we use a `for i :=` loop and `ml[i] = ml[i].With(fields...)` inside the loop?

(I'll open a separate PR to do the same throughout the core logger.)

akshayjshah commented on tee_logger_test.go:
```diff
@@ -0,0 +1,145 @@
+
+// XXX: we cannot presently write `func TestTeeLogger_Fatal(t *testing.T)`,
+// because we can't have both a spy logger and an exit stub without a
+// dependency cycle.
```
Hopefully this'll get better when/if we promote the spy logger to a full-fledged observing logger.

akshayjshah commented on test_logger_bench_test.go:
```diff
@@ -0,0 +1,68 @@
+				if cm := logger.Check(tt.lvl, tt.msg); cm.OK() {
+					cm.Write(zap.Int("i", i))
+				}
+				i = i % len(cases)
```
Should we be incrementing `i` in here?

Also, what are the benchmark results here?
